src/javascript/dimensions-metrics-explorer/components/main.tsx

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as React from "react";
+
+import { useLocalStorage, useTypedLocalStorage } from "../../hooks";
+
+import { Column } from "../common_types";
+import { CubesByColumn, cubesByColumn } from "../cubes";
+
+import SearchBox from "./searchBox";
+import ColumnGroupList from "./columnGroupList";
+
+const Main: React.FC = () => {
+  const [searchText, setSearchText] = useLocalStorage("searchText", "");
+  const [allowDeprecated, setAllowDeprecated] = useTypedLocalStorage(
+    "allowDeprecated",
+    false
+  );
+  const [onlySegments, setOnlySegments] = useTypedLocalStorage(
+    "onlySegments",
+    false
+  );
+
+  const searchTerms = React.useMemo(
+    () =>
+      searchText
+        .toLowerCase()
+        .split(/\s+/)
+        .filter(term => term.length),
+    [searchText]
+  );
+
+  // Fetch all of the columns from the metadata API
+  const [columns, setColumns] = React.useState<null | Column[]>(null);
```

Comment:
Since you're using typescript I'd recommend `<'loading' | column[]>` as your type here. Helps to indicate exactly what's going on instead of something "null" is going on.

Reply:
Stet